src/common.c

```diff
@@ -1221,21 +1221,23 @@ int stlink_core_id(stlink_t *sl) {
 int stlink_chip_id(stlink_t *sl, uint32_t *chip_id) {
     int ret;
 
-    if (sl->core_id == STM32H7_CORE_ID) {
+    uint32_t cpu_id;
+    *chip_id = 0;
+    ret = -1;
+
+    // Read the CPU ID to determine where to read the core id from
+    if (stlink_read_debug32(sl, STLINK_REG_CM3_CPUID, &cpu_id))
+        cpu_id = 0;
+
+    // If the chip is an H7, read the chipid from the new address
+    if (sl->core_id == STM32H7_CORE_ID && cpu_id == STLINK_REG_CMx_CPUID_CM7) {
         // STM32H7 chipid in 0x5c001000 (RM0433 pg3189)
         ret = stlink_read_debug32(sl, 0x5c001000, chip_id);
-    } else {
-        // default chipid address
-        ret = stlink_read_debug32(sl, 0xE0042000, chip_id);
-    }
-
-    if (ret == -1) {
-        return(ret);
     }
 
     if (*chip_id == 0) {
-        // STM32H7 chipid in 0x5c001000 (RM0433 pg3189)
-        ret = stlink_read_debug32(sl, 0x5c001000, chip_id);
+        // default chipid address
+        ret = stlink_read_debug32(sl, 0xE0042000, chip_id);
     }
 
     if (*chip_id == 0) {
```

Comment:
If you look at the resulting file, then everything is good:
https://github.com/stlink-org/stlink/blob/dcdbd61fda443c34d416d44237799a1c6e08d018/src/common.c#L1221-L1249

Reply:
Sry, but I don't see it yet. Expecting `ret` to be always returned with `stlink_read_debug32(sl, 0x40015800, chip_id)`  when `*chip_id == 0` is met. So I'm identifying this ...

```
     if (*chip_id == 0) { 
         // default chipid address 
         ret = stlink_read_debug32(sl, 0xE0042000, chip_id); 
     }
```

... as dead code...

Reply:
Line 1225: `*chip_id = 0;` Therefore this code will execute when MCU is not H7:
```c
     if (*chip_id == 0) { 
         // default chipid address 
         ret = stlink_read_debug32(sl, 0xE0042000, chip_id); 
     }
```

Reply:
I agree and see the point now, but this would then also be executed afterwards, wouldn't it?

```
    if (*chip_id == 0) {
        // Try Corex M0 DBGMCU_IDCODE register address
        ret = stlink_read_debug32(sl, 0x40015800, chip_id);
    }
```

so it should be preferable to pull that together as follows (though there is no change in functionality):

```
     if (*chip_id == 0) { 
        // Default chip_id address 
        ret = stlink_read_debug32(sl, 0xE0042000, chip_id); 

        // Try Cortex M0 DBGMCU_IDCODE register address
        ret = stlink_read_debug32(sl, 0x40015800, chip_id);
    }
```

Reply:
@Nightwalker-87 No, `chip_id` is pointer. The `stlink_read_debug32(sl, ..., chip_id);` calls change a value of the `chip_id` variable.

Reply:
Sry, I did not pay attention to that, of course - the value changes... Sry, for the noise. :see_no_evil: